dpe/src/dpe_instance.rs

```diff
@@ -144,6 +149,90 @@ impl<C: Crypto> DpeInstance<'_, C> {
+            .get_active_context_index(&cmd.handle)
+            .ok_or(DpeErrorCode::InvalidHandle)?;
+
+        // Make sure the command is coming from the right locality.
+        if context.locality != locality {
+            return Err(DpeErrorCode::InvalidHandle);
+        }
+
+        // Get TCI Nodes
+        const INITIALIZER: TciNodeData = TciNodeData::new();
+        let mut nodes = [INITIALIZER; MAX_HANDLES];
+        let tcb_count = self.get_tcb_nodes(context, &mut nodes)?;
+
+        // Hash TCI Nodes
+        let mut tci_bytes = [0u8; MAX_HANDLES * size_of::<TciNodeData>()];
+        let mut tci_offset = 0;
+        for n in &nodes[..tcb_count] {
+            tci_offset += n.serialize(&mut tci_bytes[tci_offset..])?;
+        }
+
+        let mut digest = [0; DPE_PROFILE.get_hash_size()];
+        C::hash(DPE_PROFILE, &tci_bytes[..tci_offset], &mut digest)?;
+
+        // Derive CDI and public key
+        let cdi = C::derive_cdi(DPE_PROFILE, &digest, b"DPE")?;
+        let pub_key = C::derive_ecdsa_pub(DPE_PROFILE, &cdi, &cmd.label, b"ECC")?;
+
+        // Hash the public key for the serialNumber name field
+        let mut pub_bytes = [0u8; size_of::<EcdsaPub>()];
+        let pub_offset = pub_key.serialize(&mut pub_bytes)?;
+        let mut pub_digest = [0u8; DPE_PROFILE.get_hash_size()];
+        C::hash(DPE_PROFILE, &pub_bytes[..pub_offset], &mut pub_digest)?;
+
+        // TODO: Let the platform specify issuer name
+        let issuer_name = Name {
+            cn: b"DPE Issuer",
+            serial: &[0; DPE_PROFILE.get_hash_size()],
+        };
+
+        let subject_name = Name {
+            cn: b"DPE Leaf",
+            serial: &pub_digest,
+        };
+
+        let measurements = MeasurementData {
+            _label: &cmd.label,
+            tci_nodes: &nodes[..tcb_count],
+        };
+        let mut response = CertifyKeyResp::default();
+
+        // Get certificate
+        let mut tbs_buffer = [0u8; MAX_CERT_SIZE];
+        let mut tbs_writer = X509CertWriter::new(&mut tbs_buffer);
+        let mut bytes_written = tbs_writer.encode_ecdsa_tbs(
+            /*serial=*/ &pub_digest,
+            &issuer_name,
+            &subject_name,
+            &pub_key,
+            &measurements,
+        )?;
+
+        let mut tbs_digest = [0u8; DPE_PROFILE.get_hash_size()];
+        C::hash(DPE_PROFILE, &tbs_buffer[..bytes_written], &mut tbs_digest)?;
+        let sig = C::ecdsa_sign_with_alias(DPE_PROFILE, &tbs_digest)?;
+
+        let mut cert_writer = X509CertWriter::new(&mut response.cert);
+        bytes_written = cert_writer.encode_ecdsa_certificate(&tbs_buffer[..bytes_written], &sig)?;
+        response.cert_size = match u32::try_from(bytes_written) {
+            Ok(n) => n,
+            Err(_) => return Err(DpeErrorCode::InternalError),
+        };
+
+        response.derived_pubkey_x.copy_from_slice(&pub_key.x);
+        response.derived_pubkey_y.copy_from_slice(&pub_key.y);
+
+        Ok(response)
+    }
```

Comment:
It looks like the `new_context_handle` hasn't been assigned. Alternatively, the CertifyKeyResp could be created after all of the operations to help avoid missing struct members. I don't feel too strongly either way, though. Feel free to do whatever you think is best.

```suggestion
        let mut tbs_digest = [0u8; DPE_PROFILE.get_hash_size()];
        C::hash(DPE_PROFILE, &tbs_buffer[..bytes_written], &mut tbs_digest)?;
        let sig = C::ecdsa_sign_with_alias(DPE_PROFILE, &tbs_digest)?;

        let mut cert = [0u8; MAX_CERT_SIZE];
        let mut cert_writer = X509CertWriter::new(&mut cert);
        bytes_written = cert_writer.encode_ecdsa_certificate(&tbs_buffer[..bytes_written], &sig)?;
        let cert_size = match u32::try_from(bytes_written) {
            Ok(n) => n,
            Err(_) => return Err(DpeErrorCode::InternalError),
        };

        Ok(CertifyKeyResp {
            new_context_handle: [0; HANDLE_SIZE], // would need to change this value
            derived_pubkey_x: pub_key.x,
            derived_pubkey_y: pub_key.y,
            cert_size,
            cert,
        })
```

Reply:
I think the main reason I did it this way was to avoid copying the whole cert buffer. But now that I'm thinking about it, the compiler is probably smart enough to avoid that copy.